{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cc1a895",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install chromadb sentence-transformers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "16e4c84c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sentence_transformers import SentenceTransformer\n",
    "import chromadb\n",
    "\n",
    "class GeneListRag:\n",
    "    def __init__(self, gene_list_path = '../../ref/genes.txt'):\n",
    "        self.embed_model = SentenceTransformer('all-MiniLM-L6-v2')\n",
    "        self.client = chromadb.PersistentClient(path='./gene_db')\n",
    "        self.collection = self.client.get_or_create_collection(name='genes')\n",
    "        self.gene_list_path = gene_list_path\n",
    "\n",
    "        if self.collection.count() == 0:\n",
    "            print(\"Indexing gene list...\")\n",
    "            self._index_gene_list(self.gene_list_path)\n",
    "        else :\n",
    "            print(f\"Gene list already indexed: {self.collection.count()} genes chunks found.\")\n",
    "    \n",
    "    def _index_gene_list(self, file_path):\n",
    "        # load gene list\n",
    "        with open(file_path, 'r') as f:\n",
    "            genes = f.read()\n",
    "\n",
    "        # create chunks : groups of n genes per chunk\n",
    "        genes = [gene for gene in genes.split('\\n') if gene.strip() != '']\n",
    "        chunk_size = 1\n",
    "        gene_chunks = [genes[i:i + chunk_size] for i in range(0, len(genes), chunk_size)]\n",
    "        \n",
    "        # create embeddings for each chunk and store in ChromaDB\n",
    "        for i, chunk in enumerate(gene_chunks):\n",
    "            chunk_text = '\\n'.join(chunk)\n",
    "            embedding = self.embed_model.encode(chunk_text).tolist()\n",
    "            self.collection.add(\n",
    "                documents=[chunk_text],\n",
    "                embeddings=[embedding],\n",
    "                ids=[f\"gene_chunk_{i}\"]\n",
    "            )\n",
    "        print(f\"Indexed {len(gene_chunks)} gene chunks.\")\n",
    "\n",
    "    def retrieve_relevant_genes(self, trial_criteria):\n",
    "        # create embeddings for the trial criteria\n",
    "        query_embedding = self.embed_model.encode(trial_criteria).tolist()\n",
    "\n",
    "        # query the vector database for relevant gene chunks\n",
    "        results = self.collection.query(\n",
    "            query_embeddings=[query_embedding]\n",
    "             ,n_results=5\n",
    "        )\n",
    "        print(f\"Retrieved {len(results['documents'][0])} relevant gene chunks.\")\n",
    "        retrieved_genes = []\n",
    "        for doc in results['documents'][0]:\n",
    "            print(f\"Retrieved chunk:\\n{doc}\\n\")\n",
    "            retrieved_genes.extend(doc.split('\\n')) \n",
    "        return retrieved_genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "00128674",
   "metadata": {},
   "outputs": [],
   "source": [
    "nct_id = \"NCT05410145\"\n",
    "eligibilityCriteria = \"Inclusion:\\n\\n* Subject must have a histologically or cytologically confirmed metastatic or locally advanced solid tumor which is progressing.\\n* Subject must have documented KRAS p.G12C mutation identified within the last 5 years by a local test on tumor tissue or blood.\\n* Subject must have measurable disease per RECIST v1.1.\\n* Subject must have Eastern Cooperative Oncology Group (ECOG) performance status of 0 or 1.\\n* Subject must have adequate organ and marrow function within the screening period.\\n\\nExclusion:\\n\\n* Subject has any prior treatment with other treatments without adequate washout periods as defined in the protocol.\\n* Subject has uncontrolled intercurrent illness, including but not limited to, ongoing or active infection, uncontrolled or significant cardiovascular disease, serious chronic gastrointestinal conditions associated with diarrhea, or psychiatric illness/social situations that would limit compliance with study requirements, substantially increase risk of incurring AEs, or compromise the ability of the subject to give written informed consent.\\n* Subject has unresolved treatment-related toxicities from previous anticancer therapy of NCI CTCAE Grade \\u22652 (with exception of vitiligo or alopecia).\\n* Subject has active gastrointestinal disease or other that could interfere significantly with the absorption, distribution, metabolism, or excretion of oral therapy.\\n* Concurrent participation in any clinical research study involving treatment with any investigational drug, radiotherapy, or surgery, except for the nontreatment phases of these studies (e.g., follow-up phase).\\n\\nOther protocol inclusion/exclusion criteria may apply\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "da436749",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene list already indexed: 754 genes chunks found.\n",
      "Retrieved 5 relevant gene chunks.\n",
      "Retrieved chunk:\n",
      "MED12\n",
      "\n",
      "Retrieved chunk:\n",
      "BCL11A\n",
      "\n",
      "Retrieved chunk:\n",
      "NCOR1\n",
      "\n",
      "Retrieved chunk:\n",
      "CTCF\n",
      "\n",
      "Retrieved chunk:\n",
      "TNFRSF14\n",
      "\n",
      "Retrieved gene subset: ['MED12', 'BCL11A', 'NCOR1', 'CTCF', 'TNFRSF14']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Initialize RAG system and retrieve relevant genes\n",
    "rag_system = GeneListRag()\n",
    "relevant_genes_list = rag_system.retrieve_relevant_genes(eligibilityCriteria)\n",
    "print(f'Retrieved gene subset: {relevant_genes_list}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bdad37fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2026-01-09 17:52:08.573\u001b[0m | \u001b[34m\u001b[1mDEBUG   \u001b[0m | \u001b[36mutils.ai_helper\u001b[0m:\u001b[36msend_ai_request\u001b[0m:\u001b[36m95\u001b[0m - \u001b[34m\u001b[1mAI request | ID:NCT05410145 | {\"model\": \"neuralmagic/DeepSeek-R1-Distill-Qwen-32B-quantized.w4a16\", \"messages\": [{\"role\": \"system\", \"content\": \"You are a biomedical researcher.\"}, {\"role\": \"user\", \"content\": \"Task: Evaluate the EligibilityCriteria to return a JSON-formatted eligibility criteria involving genetic variants in any genes such as those in the GeneList.\\n    EligibilityCritieria: Inclusion:\\n\\n* Subject must have a histologically or cytologically confirmed metastatic or locally advanced solid tumor which is progressing.\\n* Subject must have documented KRAS p.G12C mutation identified within the last 5 years by a local test on tumor tissue or blood.\\n* Subject must have measurable disease per RECIST v1.1.\\n* Subject must have Eastern Cooperative Oncology Group (ECOG) performance status of 0 or 1.\\n* Subject must have adequate organ and marrow function within the screening period.\\n\\nExclusion:\\n\\n* Subject has any prior treatment with other treatments without adequate washout periods as defined in the protocol.\\n* Subject has uncontrolled intercurrent illness, including but not limited to, ongoing or active infection, uncontrolled or significant cardiovascular disease, serious chronic gastrointestinal conditions associated with diarrhea, or psychiatric illness/social situations that would limit compliance with study requirements, substantially increase risk of incurring AEs, or compromise the ability of the subject to give written informed consent.\\n* Subject has unresolved treatment-related toxicities from previous anticancer therapy of NCI CTCAE Grade \\u22652 (with exception of vitiligo or alopecia).\\n* Subject has active gastrointestinal disease or other that could interfere significantly with the absorption, distribution, metabolism, or excretion of oral therapy.\\n* Concurrent participation in any clinical research study involving treatment with any investigational drug, radiotherapy, or surgery, except for the nontreatment phases of these studies (e.g., follow-up phase).\\n\\nOther protocol inclusion/exclusion criteria may apply\\n    GeneList: ['MED12', 'BCL11A', 'NCOR1', 'CTCF', 'TNFRSF14']\\n\\n    Output in JSON format such that:\\n    1. If the genes are mentioned in trial's inclusion criteria, use the \\\"or\\\" operator along with included \\\"variant_category\\\".\\n    2. If the genes are mentioned in trial's exclusion criteria, use the \\\"and\\\" operator along with the excluded \\\"variant_categorory\\\" prefixed with the \\\"!\\\" operator.\\n    3. If applicable, combine inclusion and exclusion criteria with a top level \\\"and\\\" operator.\\n    4. \\\"variant_category\\\" must be in [\\\"Mutation\\\", \\\"Copy Number Variation\\\", \\\"Structural Variation\\\", \\\"Any Variation\\\", \\\"!Mutation\\\", \\\"!Copy Number Variation\\\", \\\"!Structural Variation\\\"],\\n       where \\\"Mutation\\\" is defined narrowly to include only single nucleotide variants (SNVs) and indels.\\n    5. If a specific amino acid substitution is required, return this in the \\\"protein_change\\\" field.\\n    6. In EligibilityCriteria, the term \\\"mutant\\\" means \\\"Any Variation\\\" and \\\"non-mutant\\\" means \\\"!Any Variation\\\".\\n\\n    Example 1:\\n    Criteria: Subjects with advanced solid tumors harboring NTRK1 rearrangement or KRAS G12C will be included in this trial. \\n    Output:\\n    {\\n        \\\"or\\\": [        \\n            {\\n                \\\"genomic\\\": {\\n                    \\\"hugo_symbol\\\": \\\"NTRK1\\\",\\n                    \\\"variant_category\\\": \\\"Structural Variation\\\"\\n                }\\n            },\\n            {\\n                \\\"genomic\\\": {\\n                    \\\"hugo_symbol\\\": \\\"KRAS\\\",\\n                    \\\"variant_category\\\": \\\"Mutation\\\",\\n                    \\\"protein_change\\\": \\\"G12C\\\"\\n                }\\n            }\\n        ]\\n    }\\n\\n    Example 2:\\n    Criteria: Tumors with mutant KRAS or EGFR and wildtype BRCA1/2.\\n    Output:\\n    {\\n        \\\"and\\\":[\\n            {\\n                \\\"or\\\":[\\n                    {\\\"genomic\\\": {\\\"hugo_symbol\\\": \\\"KRAS\\\", \\\"variant_category\\\": \\\"Any Variation\\\"}},\\n                    {\\\"genomic\\\": {\\\"hugo_symbol\\\": \\\"EGFR\\\", \\\"variant_category\\\": \\\"Any Variation\\\"}}\\n                ]\\n            },\\n            {\\n                \\\"and\\\":[\\n                    {\\\"genomic\\\": {\\\"hugo_symbol\\\": \\\"BRCA1\\\", \\\"variant_category\\\": \\\"!Any Variation\\\"}},\\n                    {\\\"genomic\\\": {\\\"hugo_symbol\\\": \\\"BRCA2\\\", \\\"variant_category\\\": \\\"!Any Variation\\\"}}\\n                ]\\n\\n            }\\n        ]\\n    }\\n    \"}], \"temperature\": 0.5, \"max_tokens\": 8192, \"response_format\": {\"type\": \"json_object\"}, \"stream\": false}\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating SGLang platform...\n",
      "http://localhost:30000/v1/chat/completions\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2026-01-09 17:52:16.886\u001b[0m | \u001b[34m\u001b[1mDEBUG   \u001b[0m | \u001b[36mutils.ai_helper\u001b[0m:\u001b[36msend_ai_request\u001b[0m:\u001b[36m105\u001b[0m - \u001b[34m\u001b[1mAI response | ID:NCT05410145 | {'id': 'b1e628f3d1014273b8a523de076ddc77', 'object': 'chat.completion', 'created': 1767952336, 'model': 'neuralmagic/DeepSeek-R1-Distill-Qwen-32B-quantized.w4a16', 'choices': [{'index': 0, 'message': {'role': 'assistant', 'content': 'Okay, so I\\'m trying to figure out how to evaluate the EligibilityCriteria provided and create a JSON-formatted output based on the given GeneList. Let me break this down step by step.\\n\\nFirst, I\\'ll look at the inclusion and exclusion criteria provided. The inclusion criteria mention that the subject must have a KRAS p.G12C mutation. The exclusion criteria don\\'t mention any specific genes from the GeneList, so I don\\'t need to consider them for exclusion.\\n\\nNow, the GeneList includes [\\'MED12\\', \\'BCL11A\\', \\'NCOR1\\', \\'CTCF\\', \\'TNFRSF14\\']. I need to check if any of these genes are mentioned in the inclusion or exclusion criteria.\\n\\nLooking at the inclusion criteria, only KRAS is mentioned. Since KRAS isn\\'t in the GeneList, I don\\'t need to include it in the JSON output. The exclusion criteria don\\'t mention any of the GeneList genes either.\\n\\nSince none of the GeneList genes are mentioned in the criteria, there are no conditions to include or exclude based on these genes. Therefore, the JSON output should reflect that there are no applicable conditions.\\n\\nI think the correct approach here is to return an empty JSON object because there are no genes from the GeneList that are relevant to the eligibility criteria. This makes sense because the criteria don\\'t reference any of the specified genes, so they don\\'t affect eligibility.\\n</think>\\n\\n```json\\n{\\n    \"and\": []\\n}\\n```', 'reasoning_content': None, 'tool_calls': None}, 'logprobs': None, 'finish_reason': 'stop', 'matched_stop': 151643}], 'usage': {'prompt_tokens': 880, 'total_tokens': 1183, 'completion_tokens': 303, 'prompt_tokens_details': None}}\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Response [200]>\n",
      "200\n",
      "and: []\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import sys\n",
    "import os\n",
    "import yaml\n",
    "sys.path.append(os.path.abspath('../../'))\n",
    "from utils.ai_helper import send_ai_request, parse_ai_response, get_genomic_criteria_prompt\n",
    "\n",
    "json_schema, prompt = get_genomic_criteria_prompt(relevant_genes_list, eligibilityCriteria)\n",
    "ai_response = send_ai_request(nct_id, prompt, json_schema)\n",
    "genomic_criteria_dict = parse_ai_response(ai_response)\n",
    "\n",
    "yaml_data = yaml.dump(genomic_criteria_dict, sort_keys=False)\n",
    "print(yaml_data)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rag_test",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
