{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cc1a895",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install chromadb sentence-transformers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "16e4c84c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sentence_transformers import SentenceTransformer\n",
    "import chromadb\n",
    "import json\n",
    "\n",
    "class GeneListRag:\n",
    "    def __init__(self, gene_list_path = '../ref/genes.txt'):\n",
    "        self.embed_model = SentenceTransformer('all-MiniLM-L6-v2')\n",
    "        self.client = chromadb.PersistentClient(path='./gene_db')\n",
    "        self.collection = self.client.get_or_create_collection(name='genes')\n",
    "        self.gene_list_path = gene_list_path\n",
    "\n",
    "        if self.collection.count() == 0:\n",
    "            print(\"Indexing gene list...\")\n",
    "            self._index_gene_list(self.gene_list_path)\n",
    "        else :\n",
    "            print(f\"Gene list already indexed: {self.collection.count()} genes chunks found.\")\n",
    "    \n",
    "    def _index_gene_list(self, file_path):\n",
    "\n",
    "        # load gene list\n",
    "        with open(file_path, 'r') as f:\n",
    "            genes = f.read()\n",
    "\n",
    "        # create chunks : groups of 200 genes to keep context\n",
    "        genes = [gene for gene in genes.split('\\n') if gene.strip() != '']\n",
    "        chunk_size = 200\n",
    "        gene_chunks = [genes[i:i + chunk_size] for i in range(0, len(genes), chunk_size)]\n",
    "        \n",
    "        # create embeddings for each chunk and store in ChromaDB\n",
    "        for i, chunk in enumerate(gene_chunks):\n",
    "            chunk_text = '\\n'.join(chunk)\n",
    "            embedding = self.embed_model.encode(chunk_text).tolist()\n",
    "            self.collection.add(\n",
    "                documents=[chunk_text],\n",
    "                embeddings=[embedding],\n",
    "                ids=[f\"gene_chunk_{i}\"]\n",
    "            )\n",
    "        print(f\"Indexed {len(gene_chunks)} gene chunks.\")\n",
    "\n",
    "    def retrieve_relevant_genes(self, trial_criteria):\n",
    "        # create embeddings for the trial criteria\n",
    "        query_embedding = self.embed_model.encode(trial_criteria).tolist()\n",
    "\n",
    "        # query the vector database for relevant gene chunks\n",
    "        results = self.collection.query(\n",
    "            query_embeddings=[query_embedding],\n",
    "            n_results=5\n",
    "        )\n",
    "\n",
    "        retrieved_genes = []    \n",
    "        for doc in results['documents'][0]:\n",
    "            retrieved_genes.extend(doc.split('\\n')) \n",
    "        return retrieved_genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c9764230",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genomic_criteria_prompt(relevant_genes, eligibilityCriteria):\n",
    "    prompt = f\"\"\"\n",
    "    Task: Evaluate the clinical trial description against the provided gene list and variant categories to determine whether any mutations in the listed genes are included or excluded based on the eligibility criteria.\n",
    "\n",
    "    Instructions:\n",
    "    1. Identify if the clinical trial description mentions mutations in the given genes (inclusion) or specifies exclusions.\n",
    "    2. Use the variant categories:\n",
    "        Mutation\n",
    "        Copy Number Variation\n",
    "        Structural Variation\n",
    "        Any Variation\n",
    "        !Mutation\n",
    "        !Copy Number Variation\n",
    "        !Structural Variation\n",
    "    \n",
    "    Logic:\n",
    "    1. If the genes are mentioned in trial's inlcusion criteria, use the 'or' operator along with appropriate variant categories.\n",
    "    2. If the genes are mentioned in trial's exclusion criteria, use the 'and' operator along with variant categories (!Mutation/!Copy Number Variation/!Structural Variation).\n",
    "    3. If applicable, combine inclusion and exclusion with a top level 'and' operator.\n",
    "    \n",
    "    Gene list: {relevant_genes}\n",
    "\n",
    "    Clinical Trial Description: {eligibilityCriteria}\n",
    "\n",
    "    Example A:\n",
    "    Inclusion criteria:Subjects with advanced solid tumors harboring ROS1 or NTRK1 rearrangement will be included in this trial. \n",
    "    Output:\n",
    "{{\n",
    "    \"or\": [        \n",
    "        {{\n",
    "            \"genomic\": {{\n",
    "                \"hugo_symbol\": \"ROS1\",\n",
    "                \"variant_category\": \"Structural Variation\"\n",
    "            }}\n",
    "        }},\n",
    "        {{\n",
    "            \"genomic\": {{\n",
    "                \"hugo_symbol\": \"NTRK1\",\n",
    "                \"variant_category\": \"Structural Variation\"\n",
    "            }}\n",
    "        }}\n",
    "    ]\n",
    "}}\n",
    "\n",
    "Example B:\n",
    "Inclusion criteria:Patient should have mutation in geneA or geneB.\n",
    "Exclusion criteria: Patient should not have a mutation in geneC or geneD\n",
    "\n",
    "Output:\n",
    "{{\n",
    "    \"and\":[\n",
    "        {{\n",
    "            \"or\":[\n",
    "                {{\"genomic\": {{\"hugo_symbol\": \"geneA\",\"variant_category\": \"Mutation\"}}}},\n",
    "                {{\"genomic\": {{\"hugo_symbol\": \"geneB\",\"variant_category\": \"Mutation\"}}}}\n",
    "            ]\n",
    "        }},\n",
    "        {{\n",
    "            \"and\":[\n",
    "                {{\"genomic\": {{\"hugo_symbol\": \"geneC\",\"variant_category\": \"!Mutation\"}}}},\n",
    "                {{\"genomic\": {{\"hugo_symbol\": \"geneD\",\"variant_category\": \"!Mutation\"}}}}\n",
    "            ]\n",
    "\n",
    "        }}\n",
    "    ]\n",
    "}}\n",
    "\n",
    "    \"\"\"\n",
    "    return None, prompt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "00128674",
   "metadata": {},
   "outputs": [],
   "source": [
    "nct_id = \"\"\n",
    "eligibilityCriteria = \"Phase 1\\n\\nOral repotrectinib (TPX-0005):\\n\\nPhase 1a dose escalation, Phase 1b food-effect sub-study, and Phase 1c dose escalation with food, and Midazolam drug-drug interaction sub-study.\\n\\nPhase 2\\n\\nOral repotrectinib (TPX-0005): 6 distinct expansion cohorts\\n\\n* EXP-1: ROS1 TKI-naïve ROS1+ NSCLC\\n* EXP-2: 1 Prior ROS1 TKI and 1 Platinum based chemo ROS1+ NSCLC\\n* EXP-3: 2 Prior ROS1 TKIs ROS1+ NSCLC (No Chemo or IO)\\n* EXP-4: 1 Prior ROS1 TKI ROS1+ NSCLC (No Chemo or IO)\\n* EXP-5: TRK TKI-naïve NTRK+ solid tumors\\n* EXP-6: TRK TKI-pretreated NTRK+ solid tumors\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "da436749",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2026-01-02 20:29:41.830\u001b[0m | \u001b[34m\u001b[1mDEBUG   \u001b[0m | \u001b[36mutils.ai_helper\u001b[0m:\u001b[36msend_ai_request\u001b[0m:\u001b[36m112\u001b[0m - \u001b[34m\u001b[1mAI request | ID: | {\"model\": \"hf.co/unsloth/gemma-3-27b-it-GGUF:Q4_K_M\", \"messages\": [{\"role\": \"system\", \"content\": \"You are a biomedical researcher specializing in cancer genomics and clinical trials.\"}, {\"role\": \"user\", \"content\": \"\\n    Task: Evaluate the clinical trial description against the provided gene list and variant categories to determine whether any mutations in the listed genes are included or excluded based on the eligibility criteria.\\n\\n    Instructions:\\n    1. Identify if the clinical trial description mentions mutations in the given genes (inclusion) or specifies exclusions.\\n    2. Use the variant categories:\\n        Mutation\\n        Copy Number Variation\\n        Structural Variation\\n        Any Variation\\n        !Mutation\\n        !Copy Number Variation\\n        !Structural Variation\\n\\n    Logic:\\n    1. If the genes are mentioned in trial's inlcusion criteria, use the 'or' operator along with appropriate variant categories.\\n    2. If the genes are mentioned in trial's exclusion criteria, use the 'and' operator along with variant categories (!Mutation/!Copy Number Variation/!Structural Variation).\\n    3. If applicable, combine inclusion and exclusion with a top level 'and' operator.\\n\\n    Gene list: ['A1CF', 'ABI1', 'ABL1', 'ABL2', 'ACKR3', 'ACSL3', 'ACSL6', 'ACVR1', 'ACVR1B', 'ACVR2A', 'AFDN', 'AFF1', 'AFF3', 'AFF4', 'AKAP9', 'AKT1', 'AKT2', 'AKT3', 'ALDH2', 'ALK', 'AMER1', 'ANK1', 'APC', 'APOBEC3B', 'AR', 'ARAF', 'ARHGAP26', 'ARHGAP35', 'ARHGAP5', 'ARHGEF10', 'ARHGEF10L', 'ARHGEF12', 'ARID1A', 'ARID1B', 'ARID2', 'ARNT', 'ASPM', 'ASPSCR1', 'ASXL1', 'ASXL2', 'ATF1', 'ATIC', 'ATM', 'ATP1A1', 'ATP2B3', 'ATR', 'ATRX', 'AXIN1', 'AXIN2', 'B2M', 'BAP1', 'BARD1', 'BAX', 'BAZ1A', 'BCL10', 'BCL11A', 'BCL11B', 'BCL2', 'BCL2L12', 'BCL3', 'BCL6', 'BCL7A', 'BCL9', 'BCL9L', 'BCLAF1', 'BCOR', 'BCORL1', 'BCR', 'BIRC3', 'BIRC6', 'BLM', 'BMP5', 'BMPR1A', 'BRAF', 'BRCA1', 'BRCA2', 'BRD3', 'BRD4', 'BRIP1', 'BTG1', 'BTG2', 'BTK', 'BUB1B', 'C15orf65', 'CACNA1D', 'CALR', 'CAMTA1', 'CANT1', 'CARD11', 'CARS', 'CASP3', 'CASP8', 'CASP9', 'CBFA2T3', 'CBFB', 'CBL', 'CBLB', 'CBLC', 'CCDC6', 'CCNB1IP1', 'CCNC', 'CCND1', 'CCND2', 'CCND3', 'CCNE1', 'CCR4', 'CCR7', 'CD209', 'CD274', 'CD28', 'CD74', 'CD79A', 'CD79B', 'CDC73', 'CDH1', 'CDH10', 'CDH11', 'CDH17', 'CDK12', 'CDK4', 'CDK6', 'CDKN1A', 'CDKN1B', 'CDKN2A', 'CDKN2C', 'CDX2', 'CEBPA', 'CEP89', 'CHCHD7', 'CHD2', 'CHD4', 'CHEK2', 'CHIC2', 'CHST11', 'CIC', 'CIITA', 'CLIP1', 'CLP1', 'CLTC', 'CLTCL1', 'CNBD1', 'CNBP', 'CNOT3', 'CNTNAP2', 'CNTRL', 'COL1A1', 'COL2A1', 'COL3A1', 'COX6C', 'CPEB3', 'CREB1', 'CREB3L1', 'CREB3L2', 'CREBBP', 'CRLF2', 'CRNKL1', 'CRTC1', 'CRTC3', 'CSF1R', 'CSF3R', 'CSMD3', 'CTCF', 'CTNNA1', 'CTNNA2', 'CTNNB1', 'CTNND1', 'CTNND2', 'CUL3', 'CUX1', 'CXCR4', 'CYLD', 'CYP2C8', 'CYSLTR2', 'DAXX', 'DCAF12L2', 'DCC', 'DCTN1', 'DDB2', 'DDIT3', 'DDR2', 'DDX10', 'DDX3X', 'DDX5', 'DDX6', 'DEK', 'DGCR8', 'DICER1', 'DNAJB1', 'DNM2', 'DNMT3A', 'DROSHA', 'EBF1', 'ECT2L', 'EED', 'EGFR', 'EIF1AX', 'EIF3E', 'EIF4A2', 'ELF3', 'ELF4', 'MAP3K13', 'MAPK1', 'MAX', 'MB21D2', 'MDM2', 'MDM4', 'MDS2', 'MECOM', 'MED12', 'MEN1', 'MET', 'MGMT', 'MITF', 'MLF1', 'MLH1', 'MLLT1', 'MLLT10', 'MLLT11', 'MLLT3', 'MLLT6', 'MN1', 'MNX1', 'MPL', 'MRTFA', 'MSH2', 'MSH6', 'MSI2', 'MSN', 'MTCP1', 'MTOR', 'MUC1', 'MUC16', 'MUC4', 'MUC6', 'MUTYH', 'MYB', 'MYC', 'MYCL', 'MYCN', 'MYD88', 'MYH11', 'MYH9', 'MYO5A', 'MYOD1', 'N4BP2', 'NAB2', 'NACA', 'NBEA', 'NBN', 'NCKIPSD', 'NCOA1', 'NCOA2', 'NCOA4', 'NCOR1', 'NCOR2', 'NDRG1', 'NF1', 'NF2', 'NFATC2', 'NFE2L2', 'NFIB', 'NFKB2', 'NFKBIE', 'NIN', 'NKX2-1', 'NONO', 'NOTCH1', 'NOTCH2', 'NPM1', 'NR2F2', 'NR4A3', 'NRAS', 'NRG1', 'NSD1', 'NSD2', 'NSD3', 'NT5C2', 'NTHL1', 'NTRK1', 'NTRK2', 'NTRK3', 'NUMA1', 'NUP214', 'NUP98', 'NUTM1', 'NUTM2B', 'NUTM2D', 'OLIG2', 'OMD', 'P2RY8', 'PABPC1', 'PAFAH1B2', 'PALB2', 'PATZ1', 'PAX3', 'PAX5', 'PAX7', 'PAX8', 'PBRM1', 'PBX1', 'PCBP1', 'PCM1', 'PDCD1LG2', 'PDE4DIP', 'PDGFB', 'PDGFRA', 'PDGFRB', 'PER1', 'PHF6', 'PHOX2B', 'PICALM', 'PIK3CA', 'PIK3CB', 'PIK3R1', 'PIK3R2', 'PIK3R3', 'PIM1', 'PLAG1', 'PLCG1', 'PLEC', 'PML', 'PMS1', 'PMS2', 'POLD1', 'POLE', 'POLG', 'POLQ', 'POLR2A', 'POT1', 'POU2AF1', 'POU5F1', 'PPARG', 'PPFIBP1', 'PPM1D', 'PPP2R1A', 'PPP6C', 'PRCC', 'PRDM1', 'PRDM16', 'PRDM2', 'PREX2', 'PRF1', 'PRKACA', 'PRKAR1A', 'PRKCB', 'PRKD1', 'PRPF40B', 'PRRX1', 'PSIP1', 'PTCH1', 'PTEN', 'PTK6', 'PTPN11', 'PTPN13', 'PTPN6', 'PTPRB', 'PTPRC', 'PTPRD', 'PTPRK', 'PTPRT', 'PWWP2A', 'QKI', 'RABEP1', 'RAC1', 'RAD17', 'RAD21', 'RAD50', 'RAD51B', 'RAD51C', 'RAF1', 'RALGDS', 'RANBP2', 'RAP1B', 'RAP1GDS1', 'RARA', 'RB1', 'RBM10', 'RBM15', 'RECQL4', 'REL', 'RET', 'RFWD3', 'RGPD3', 'RGS7', 'RHOA', 'RHOH', 'RMI2', 'RNF213', 'RNF43', 'ROBO2', 'ROS1', 'RPL10', 'RPL22', 'RPL5', 'RPN1', 'RRAS2', 'RSPO2', 'RSPO3', 'RUNX1', 'RUNX1T1', 'RXRA', 'S100A7', 'SALL4', 'SBDS', 'SDC4', 'SDHA', 'SDHAF2', 'SDHB', 'SDHC', 'SDHD', 'SEPT5', 'SEPT6', 'SEPT9', 'SET', 'SETBP1', 'SETD1B', 'SETD2', 'SETDB1', 'SF3B1', 'SFPQ', 'SFRP4', 'SGK1', 'SH2B3', 'SH3GL1', 'SHTN1', 'SIRPA', 'SIX1', 'SIX2', 'SKI', 'SLC34A2', 'SLC45A3', 'SMAD2', 'SMAD3', 'SMAD4', 'SMARCA4', 'SMARCB1', 'SMARCD1', 'SMARCE1', 'SMC1A', 'SMO', 'SND1', 'SNX29', 'SOCS1', 'SOS1', 'SOX2', 'SOX21', 'SPECC1', 'SPEN', 'SPOP', 'SRC', 'SRGAP3', 'SRSF2', 'SRSF3', 'SS18', 'SS18L1', 'SSX1', 'SSX2', 'SSX4', 'STAG1', 'STAG2', 'STAT3', 'STAT5B', 'STAT6', 'STIL', 'STK11', 'STRN', 'SUB1', 'SUFU', 'SUZ12', 'SYK', 'TAF15', 'TAL1', 'TAL2', 'TBL1XR1', 'TBX3', 'TCEA1', 'TCF12', 'TCF3', 'TCF7L2', 'TCL1A', 'TEC', 'TENT5C', 'TERT', 'TET1', 'TET2', 'TFE3', 'TFEB', 'TFG', 'TFPT', 'TFRC', 'TGFBR2', 'THRAP3', 'TLX1', 'TLX3', 'TMEM127', 'TMPRSS2', 'TMSB4X', 'TNC', 'TNFAIP3', 'TNFRSF14', 'TNFRSF17', 'TNRC18', 'TOP1', 'TP53', 'TP63', 'TPM3', 'TPM4', 'TPR', 'TRA', 'TRAF7', 'TRB', 'TRD', 'TRIM24', 'TRIM27', 'TRIM33', 'TRIP11', 'TRRAP', 'TSC1', 'TSC2', 'TSHR', 'U2AF1', 'UBR5', 'USP44', 'USP6', 'USP8', 'USP9X', 'VAV1', 'VHL', 'VTI1A', 'WAS', 'WDCP', 'WIF1', 'WNK2', 'WRN', 'WT1', 'WWTR1', 'XPA', 'XPC', 'XPO1', 'YAP1', 'YWHAE', 'ZBTB16', 'ZCCHC8', 'ZEB1', 'ZFHX3', 'ZMYM2', 'ZMYM3', 'ZNF331', 'ZNF384', 'ZNF429', 'ZNF479', 'ZNF521', 'ZNRF3', 'ZRSR2', 'ELK4', 'ELL', 'ELN', 'EML4', 'EP300', 'EPAS1', 'EPHA3', 'EPHA7', 'EPS15', 'ERBB2', 'ERBB3', 'ERBB4', 'ERC1', 'ERCC2', 'ERCC3', 'ERCC4', 'ERCC5', 'ERG', 'ESR1', 'ETNK1', 'ETV1', 'ETV4', 'ETV5', 'ETV6', 'EWSR1', 'EXT1', 'EXT2', 'EZH2', 'EZR', 'FADD', 'FAM131B', 'FAM135B', 'FAM47C', 'FANCA', 'FANCC', 'FANCD2', 'FANCE', 'FANCF', 'FANCG', 'FAS', 'FAT1', 'FAT3', 'FAT4', 'FBLN2', 'FBXO11', 'FBXW7', 'FCGR2B', 'FCRL4', 'FEN1', 'FES', 'FEV', 'FGFR1', 'FGFR1OP', 'FGFR2', 'FGFR3', 'FGFR4', 'FH', 'FHIT', 'FIP1L1', 'FKBP9', 'FLCN', 'FLI1', 'FLNA', 'FLT3', 'FLT4', 'FNBP1', 'FOXA1', 'FOXL2', 'FOXO1', 'FOXO3', 'FOXO4', 'FOXP1', 'FOXR1', 'FSTL3', 'FUBP1', 'FUS', 'GAS7', 'GATA1', 'GATA2', 'GATA3', 'GLI1', 'GMPS', 'GNA11', 'GNAQ', 'GNAS', 'GOLGA5', 'GOLPH3', 'GOPC', 'GPC3', 'GPC5', 'GPHN', 'GRIN2A', 'GRM3', 'GSK3B', 'GTF2I', 'H3F3A', 'H3F3B', 'HER2', 'HERPUD1', 'HEY1', 'HGF', 'HIF1A', 'HIP1', 'HIST1H3B', 'HIST1H4I', 'HLA-A', 'HLF', 'HMGA1', 'HMGA2', 'HMGN2P46', 'HNF1A', 'HNRNPA2B1', 'HOOK3', 'HOXA11', 'HOXA13', 'HOXA9', 'HOXC11', 'HOXC13', 'HOXD11', 'HOXD13', 'HRAS', 'HSP90AA1', 'HSP90AB1', 'ID3', 'IDH1', 'IDH2', 'IGF2BP2', 'IGH', 'IGK', 'IGL', 'IKBKB', 'IKZF1', 'IKZF3', 'IL2', 'IL21R', 'IL6ST', 'IL7R', 'IRF4', 'IRS4', 'ISX', 'ITGAV', 'ITK', 'JAK1', 'JAK2', 'JAK3', 'JAZF1', 'JUN', 'KAT6A', 'KAT6B', 'KAT7', 'KCNJ5', 'KDM5A', 'KDM5C', 'KDM6A', 'KDR', 'KDSR', 'KEAP1', 'KIAA1549', 'KIF5B', 'KIT', 'KLF4', 'KLF6', 'KLK2', 'KMT2A', 'KMT2C', 'KMT2D', 'KNL1', 'KNSTRN', 'KRAS', 'KTN1', 'LARP4B', 'LASP1', 'LATS1', 'LATS2', 'LCK', 'LCP1', 'LEF1', 'LEPROTL1', 'LHFPL6', 'LIFR', 'LMNA', 'LMO1', 'LMO2', 'LPP', 'LRIG3', 'LRP1B', 'LSM14A', 'LYL1', 'LYN', 'LZTR1', 'MACC1', 'MAF', 'MAFB', 'MALAT1', 'MALT1', 'MAML2', 'MAP2K1', 'MAP2K2', 'MAP2K4', 'MAP3K1']\\n\\n    Clinical Trial Description: Phase 1\\n\\nOral repotrectinib (TPX-0005):\\n\\nPhase 1a dose escalation, Phase 1b food-effect sub-study, and Phase 1c dose escalation with food, and Midazolam drug-drug interaction sub-study.\\n\\nPhase 2\\n\\nOral repotrectinib (TPX-0005): 6 distinct expansion cohorts\\n\\n* EXP-1: ROS1 TKI-na\\u00efve ROS1+ NSCLC\\n* EXP-2: 1 Prior ROS1 TKI and 1 Platinum based chemo ROS1+ NSCLC\\n* EXP-3: 2 Prior ROS1 TKIs ROS1+ NSCLC (No Chemo or IO)\\n* EXP-4: 1 Prior ROS1 TKI ROS1+ NSCLC (No Chemo or IO)\\n* EXP-5: TRK TKI-na\\u00efve NTRK+ solid tumors\\n* EXP-6: TRK TKI-pretreated NTRK+ solid tumors\\n\\n    Example A:\\n    Inclusion criteria:Subjects with advanced solid tumors harboring ROS1 or NTRK1 rearrangement will be included in this trial. \\n    Output:\\n{\\n    \\\"or\\\": [        \\n        {\\n            \\\"genomic\\\": {\\n                \\\"hugo_symbol\\\": \\\"ROS1\\\",\\n                \\\"variant_category\\\": \\\"Structural Variation\\\"\\n            }\\n        },\\n        {\\n            \\\"genomic\\\": {\\n                \\\"hugo_symbol\\\": \\\"NTRK1\\\",\\n                \\\"variant_category\\\": \\\"Structural Variation\\\"\\n            }\\n        }\\n    ]\\n}\\n\\nExample B:\\nInclusion criteria:Patient should have mutation in geneA or geneB.\\nExclusion criteria: Patient should not have a mutation in geneC or geneD\\n\\nOutput:\\n{\\n    \\\"and\\\":[\\n        {\\n            \\\"or\\\":[\\n                {\\\"genomic\\\": {\\\"hugo_symbol\\\": \\\"geneA\\\",\\\"variant_category\\\": \\\"Mutation\\\"}},\\n                {\\\"genomic\\\": {\\\"hugo_symbol\\\": \\\"geneB\\\",\\\"variant_category\\\": \\\"Mutation\\\"}}\\n            ]\\n        },\\n        {\\n            \\\"and\\\":[\\n                {\\\"genomic\\\": {\\\"hugo_symbol\\\": \\\"geneC\\\",\\\"variant_category\\\": \\\"!Mutation\\\"}},\\n                {\\\"genomic\\\": {\\\"hugo_symbol\\\": \\\"geneD\\\",\\\"variant_category\\\": \\\"!Mutation\\\"}}\\n            ]\\n\\n        }\\n    ]\\n}\\n\\n    \"}], \"stream\": false, \"options\": {\"think\": true, \"temperature\": 0, \"seed\": 42, \"top_k\": 1}, \"format\": \"json\"}\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene list already indexed: 4 genes chunks found.\n",
      "Retrieved gene subset: ['A1CF', 'ABI1', 'ABL1', 'ABL2', 'ACKR3', 'ACSL3', 'ACSL6', 'ACVR1', 'ACVR1B', 'ACVR2A', 'AFDN', 'AFF1', 'AFF3', 'AFF4', 'AKAP9', 'AKT1', 'AKT2', 'AKT3', 'ALDH2', 'ALK', 'AMER1', 'ANK1', 'APC', 'APOBEC3B', 'AR', 'ARAF', 'ARHGAP26', 'ARHGAP35', 'ARHGAP5', 'ARHGEF10', 'ARHGEF10L', 'ARHGEF12', 'ARID1A', 'ARID1B', 'ARID2', 'ARNT', 'ASPM', 'ASPSCR1', 'ASXL1', 'ASXL2', 'ATF1', 'ATIC', 'ATM', 'ATP1A1', 'ATP2B3', 'ATR', 'ATRX', 'AXIN1', 'AXIN2', 'B2M', 'BAP1', 'BARD1', 'BAX', 'BAZ1A', 'BCL10', 'BCL11A', 'BCL11B', 'BCL2', 'BCL2L12', 'BCL3', 'BCL6', 'BCL7A', 'BCL9', 'BCL9L', 'BCLAF1', 'BCOR', 'BCORL1', 'BCR', 'BIRC3', 'BIRC6', 'BLM', 'BMP5', 'BMPR1A', 'BRAF', 'BRCA1', 'BRCA2', 'BRD3', 'BRD4', 'BRIP1', 'BTG1', 'BTG2', 'BTK', 'BUB1B', 'C15orf65', 'CACNA1D', 'CALR', 'CAMTA1', 'CANT1', 'CARD11', 'CARS', 'CASP3', 'CASP8', 'CASP9', 'CBFA2T3', 'CBFB', 'CBL', 'CBLB', 'CBLC', 'CCDC6', 'CCNB1IP1', 'CCNC', 'CCND1', 'CCND2', 'CCND3', 'CCNE1', 'CCR4', 'CCR7', 'CD209', 'CD274', 'CD28', 'CD74', 'CD79A', 'CD79B', 'CDC73', 'CDH1', 'CDH10', 'CDH11', 'CDH17', 'CDK12', 'CDK4', 'CDK6', 'CDKN1A', 'CDKN1B', 'CDKN2A', 'CDKN2C', 'CDX2', 'CEBPA', 'CEP89', 'CHCHD7', 'CHD2', 'CHD4', 'CHEK2', 'CHIC2', 'CHST11', 'CIC', 'CIITA', 'CLIP1', 'CLP1', 'CLTC', 'CLTCL1', 'CNBD1', 'CNBP', 'CNOT3', 'CNTNAP2', 'CNTRL', 'COL1A1', 'COL2A1', 'COL3A1', 'COX6C', 'CPEB3', 'CREB1', 'CREB3L1', 'CREB3L2', 'CREBBP', 'CRLF2', 'CRNKL1', 'CRTC1', 'CRTC3', 'CSF1R', 'CSF3R', 'CSMD3', 'CTCF', 'CTNNA1', 'CTNNA2', 'CTNNB1', 'CTNND1', 'CTNND2', 'CUL3', 'CUX1', 'CXCR4', 'CYLD', 'CYP2C8', 'CYSLTR2', 'DAXX', 'DCAF12L2', 'DCC', 'DCTN1', 'DDB2', 'DDIT3', 'DDR2', 'DDX10', 'DDX3X', 'DDX5', 'DDX6', 'DEK', 'DGCR8', 'DICER1', 'DNAJB1', 'DNM2', 'DNMT3A', 'DROSHA', 'EBF1', 'ECT2L', 'EED', 'EGFR', 'EIF1AX', 'EIF3E', 'EIF4A2', 'ELF3', 'ELF4', 'MAP3K13', 'MAPK1', 'MAX', 'MB21D2', 'MDM2', 'MDM4', 'MDS2', 'MECOM', 'MED12', 'MEN1', 'MET', 'MGMT', 'MITF', 'MLF1', 'MLH1', 'MLLT1', 'MLLT10', 'MLLT11', 'MLLT3', 'MLLT6', 'MN1', 'MNX1', 'MPL', 'MRTFA', 'MSH2', 'MSH6', 'MSI2', 'MSN', 'MTCP1', 'MTOR', 'MUC1', 'MUC16', 'MUC4', 'MUC6', 'MUTYH', 'MYB', 'MYC', 'MYCL', 'MYCN', 'MYD88', 'MYH11', 'MYH9', 'MYO5A', 'MYOD1', 'N4BP2', 'NAB2', 'NACA', 'NBEA', 'NBN', 'NCKIPSD', 'NCOA1', 'NCOA2', 'NCOA4', 'NCOR1', 'NCOR2', 'NDRG1', 'NF1', 'NF2', 'NFATC2', 'NFE2L2', 'NFIB', 'NFKB2', 'NFKBIE', 'NIN', 'NKX2-1', 'NONO', 'NOTCH1', 'NOTCH2', 'NPM1', 'NR2F2', 'NR4A3', 'NRAS', 'NRG1', 'NSD1', 'NSD2', 'NSD3', 'NT5C2', 'NTHL1', 'NTRK1', 'NTRK2', 'NTRK3', 'NUMA1', 'NUP214', 'NUP98', 'NUTM1', 'NUTM2B', 'NUTM2D', 'OLIG2', 'OMD', 'P2RY8', 'PABPC1', 'PAFAH1B2', 'PALB2', 'PATZ1', 'PAX3', 'PAX5', 'PAX7', 'PAX8', 'PBRM1', 'PBX1', 'PCBP1', 'PCM1', 'PDCD1LG2', 'PDE4DIP', 'PDGFB', 'PDGFRA', 'PDGFRB', 'PER1', 'PHF6', 'PHOX2B', 'PICALM', 'PIK3CA', 'PIK3CB', 'PIK3R1', 'PIK3R2', 'PIK3R3', 'PIM1', 'PLAG1', 'PLCG1', 'PLEC', 'PML', 'PMS1', 'PMS2', 'POLD1', 'POLE', 'POLG', 'POLQ', 'POLR2A', 'POT1', 'POU2AF1', 'POU5F1', 'PPARG', 'PPFIBP1', 'PPM1D', 'PPP2R1A', 'PPP6C', 'PRCC', 'PRDM1', 'PRDM16', 'PRDM2', 'PREX2', 'PRF1', 'PRKACA', 'PRKAR1A', 'PRKCB', 'PRKD1', 'PRPF40B', 'PRRX1', 'PSIP1', 'PTCH1', 'PTEN', 'PTK6', 'PTPN11', 'PTPN13', 'PTPN6', 'PTPRB', 'PTPRC', 'PTPRD', 'PTPRK', 'PTPRT', 'PWWP2A', 'QKI', 'RABEP1', 'RAC1', 'RAD17', 'RAD21', 'RAD50', 'RAD51B', 'RAD51C', 'RAF1', 'RALGDS', 'RANBP2', 'RAP1B', 'RAP1GDS1', 'RARA', 'RB1', 'RBM10', 'RBM15', 'RECQL4', 'REL', 'RET', 'RFWD3', 'RGPD3', 'RGS7', 'RHOA', 'RHOH', 'RMI2', 'RNF213', 'RNF43', 'ROBO2', 'ROS1', 'RPL10', 'RPL22', 'RPL5', 'RPN1', 'RRAS2', 'RSPO2', 'RSPO3', 'RUNX1', 'RUNX1T1', 'RXRA', 'S100A7', 'SALL4', 'SBDS', 'SDC4', 'SDHA', 'SDHAF2', 'SDHB', 'SDHC', 'SDHD', 'SEPT5', 'SEPT6', 'SEPT9', 'SET', 'SETBP1', 'SETD1B', 'SETD2', 'SETDB1', 'SF3B1', 'SFPQ', 'SFRP4', 'SGK1', 'SH2B3', 'SH3GL1', 'SHTN1', 'SIRPA', 'SIX1', 'SIX2', 'SKI', 'SLC34A2', 'SLC45A3', 'SMAD2', 'SMAD3', 'SMAD4', 'SMARCA4', 'SMARCB1', 'SMARCD1', 'SMARCE1', 'SMC1A', 'SMO', 'SND1', 'SNX29', 'SOCS1', 'SOS1', 'SOX2', 'SOX21', 'SPECC1', 'SPEN', 'SPOP', 'SRC', 'SRGAP3', 'SRSF2', 'SRSF3', 'SS18', 'SS18L1', 'SSX1', 'SSX2', 'SSX4', 'STAG1', 'STAG2', 'STAT3', 'STAT5B', 'STAT6', 'STIL', 'STK11', 'STRN', 'SUB1', 'SUFU', 'SUZ12', 'SYK', 'TAF15', 'TAL1', 'TAL2', 'TBL1XR1', 'TBX3', 'TCEA1', 'TCF12', 'TCF3', 'TCF7L2', 'TCL1A', 'TEC', 'TENT5C', 'TERT', 'TET1', 'TET2', 'TFE3', 'TFEB', 'TFG', 'TFPT', 'TFRC', 'TGFBR2', 'THRAP3', 'TLX1', 'TLX3', 'TMEM127', 'TMPRSS2', 'TMSB4X', 'TNC', 'TNFAIP3', 'TNFRSF14', 'TNFRSF17', 'TNRC18', 'TOP1', 'TP53', 'TP63', 'TPM3', 'TPM4', 'TPR', 'TRA', 'TRAF7', 'TRB', 'TRD', 'TRIM24', 'TRIM27', 'TRIM33', 'TRIP11', 'TRRAP', 'TSC1', 'TSC2', 'TSHR', 'U2AF1', 'UBR5', 'USP44', 'USP6', 'USP8', 'USP9X', 'VAV1', 'VHL', 'VTI1A', 'WAS', 'WDCP', 'WIF1', 'WNK2', 'WRN', 'WT1', 'WWTR1', 'XPA', 'XPC', 'XPO1', 'YAP1', 'YWHAE', 'ZBTB16', 'ZCCHC8', 'ZEB1', 'ZFHX3', 'ZMYM2', 'ZMYM3', 'ZNF331', 'ZNF384', 'ZNF429', 'ZNF479', 'ZNF521', 'ZNRF3', 'ZRSR2', 'ELK4', 'ELL', 'ELN', 'EML4', 'EP300', 'EPAS1', 'EPHA3', 'EPHA7', 'EPS15', 'ERBB2', 'ERBB3', 'ERBB4', 'ERC1', 'ERCC2', 'ERCC3', 'ERCC4', 'ERCC5', 'ERG', 'ESR1', 'ETNK1', 'ETV1', 'ETV4', 'ETV5', 'ETV6', 'EWSR1', 'EXT1', 'EXT2', 'EZH2', 'EZR', 'FADD', 'FAM131B', 'FAM135B', 'FAM47C', 'FANCA', 'FANCC', 'FANCD2', 'FANCE', 'FANCF', 'FANCG', 'FAS', 'FAT1', 'FAT3', 'FAT4', 'FBLN2', 'FBXO11', 'FBXW7', 'FCGR2B', 'FCRL4', 'FEN1', 'FES', 'FEV', 'FGFR1', 'FGFR1OP', 'FGFR2', 'FGFR3', 'FGFR4', 'FH', 'FHIT', 'FIP1L1', 'FKBP9', 'FLCN', 'FLI1', 'FLNA', 'FLT3', 'FLT4', 'FNBP1', 'FOXA1', 'FOXL2', 'FOXO1', 'FOXO3', 'FOXO4', 'FOXP1', 'FOXR1', 'FSTL3', 'FUBP1', 'FUS', 'GAS7', 'GATA1', 'GATA2', 'GATA3', 'GLI1', 'GMPS', 'GNA11', 'GNAQ', 'GNAS', 'GOLGA5', 'GOLPH3', 'GOPC', 'GPC3', 'GPC5', 'GPHN', 'GRIN2A', 'GRM3', 'GSK3B', 'GTF2I', 'H3F3A', 'H3F3B', 'HER2', 'HERPUD1', 'HEY1', 'HGF', 'HIF1A', 'HIP1', 'HIST1H3B', 'HIST1H4I', 'HLA-A', 'HLF', 'HMGA1', 'HMGA2', 'HMGN2P46', 'HNF1A', 'HNRNPA2B1', 'HOOK3', 'HOXA11', 'HOXA13', 'HOXA9', 'HOXC11', 'HOXC13', 'HOXD11', 'HOXD13', 'HRAS', 'HSP90AA1', 'HSP90AB1', 'ID3', 'IDH1', 'IDH2', 'IGF2BP2', 'IGH', 'IGK', 'IGL', 'IKBKB', 'IKZF1', 'IKZF3', 'IL2', 'IL21R', 'IL6ST', 'IL7R', 'IRF4', 'IRS4', 'ISX', 'ITGAV', 'ITK', 'JAK1', 'JAK2', 'JAK3', 'JAZF1', 'JUN', 'KAT6A', 'KAT6B', 'KAT7', 'KCNJ5', 'KDM5A', 'KDM5C', 'KDM6A', 'KDR', 'KDSR', 'KEAP1', 'KIAA1549', 'KIF5B', 'KIT', 'KLF4', 'KLF6', 'KLK2', 'KMT2A', 'KMT2C', 'KMT2D', 'KNL1', 'KNSTRN', 'KRAS', 'KTN1', 'LARP4B', 'LASP1', 'LATS1', 'LATS2', 'LCK', 'LCP1', 'LEF1', 'LEPROTL1', 'LHFPL6', 'LIFR', 'LMNA', 'LMO1', 'LMO2', 'LPP', 'LRIG3', 'LRP1B', 'LSM14A', 'LYL1', 'LYN', 'LZTR1', 'MACC1', 'MAF', 'MAFB', 'MALAT1', 'MALT1', 'MAML2', 'MAP2K1', 'MAP2K2', 'MAP2K4', 'MAP3K1']\n",
      "http://localhost:11434/api/chat\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m2026-01-02 20:29:47.498\u001b[0m | \u001b[34m\u001b[1mDEBUG   \u001b[0m | \u001b[36mutils.ai_helper\u001b[0m:\u001b[36msend_ai_request\u001b[0m:\u001b[36m122\u001b[0m - \u001b[34m\u001b[1mAI response | ID: | {'model': 'hf.co/unsloth/gemma-3-27b-it-GGUF:Q4_K_M', 'created_at': '2026-01-02T12:29:47.496890941Z', 'message': {'role': 'assistant', 'content': '{\"or\": [\\n    {\\n        \"genomic\": {\\n            \"hugo_symbol\": \"ROS1\",\\n            \"variant_category\": \"Structural Variation\"\\n        }\\n    },\\n    {\\n        \"genomic\": {\\n            \"hugo_symbol\": \"NTRK1\",\\n            \"variant_category\": \"Structural Variation\"\\n        }\\n    }\\n]}'}, 'done': True, 'done_reason': 'stop', 'total_duration': 5663094117, 'load_duration': 311912864, 'prompt_eval_count': 4096, 'prompt_eval_duration': 2878226827, 'eval_count': 87, 'eval_duration': 2199265013}\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Response [200]>\n",
      "200\n",
      "{'or': [{'genomic': {'hugo_symbol': 'ROS1',\n",
      "                     'variant_category': 'Structural Variation'}},\n",
      "        {'genomic': {'hugo_symbol': 'NTRK1',\n",
      "                     'variant_category': 'Structural Variation'}}]}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import sys\n",
    "import os\n",
    "\n",
    "\n",
    "sys.path.append(os.path.abspath('../'))\n",
    "\n",
    "from utils.ai_helper import send_ai_request, parse_ai_response\n",
    "\n",
    "rag_system = GeneListRag()\n",
    "relevant_genes_str = rag_system.retrieve_relevant_genes(eligibilityCriteria)\n",
    "print(f'Retrieved gene subset: {relevant_genes_str}')\n",
    "\n",
    "#print(f\"Retrieved gene subset: {relevant_genes_list}...\")\n",
    "\n",
    "json_schema, prompt = get_genomic_criteria_prompt(relevant_genes_str, eligibilityCriteria)\n",
    "ai_response = send_ai_request(nct_id, prompt, json_schema)\n",
    "genomic_criteria_dict = parse_ai_response(ai_response)\n",
    "\n",
    "# pretty print the genomic criteria dict\n",
    "import pprint\n",
    "pprint.pprint(genomic_criteria_dict)\n",
    "        "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rag_test",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
